### ISSUE TITLE
key error in Tm_NN calculation for melting temperature

### ISSUE THREAD

**Issue by jlistgarten** [biopython/biopython#682]
For this e.g., when I use shift=0, it is fine, but with any other value of shift (e.g. shift=1, shift=-1) I get an error I can't understand how to fix:

import Bio.SeqUtils.MeltingTemp as Tm
Tm.Tm_NN('AAAGCAGCCTAGGACAGCAGT','TTTCGTCGGTCCTGTCGTCA', shift=1)

The error is: 
*** KeyError: '.A/TG'

Any tips?

**Comment by peterjc** [comment]
Note you are calling the function wrongly. The signature is:

```
def Tm_NN(seq, check=True, strict=True, c_seq=None, shift=0, nn_table=DNA_NN3,
          tmm_table=DNA_TMM1, imm_table=DNA_IMM1, de_table=DNA_DE1,
          dnac1=25, dnac2=25, selfcomp=False, Na=50, K=0, Tris=0, Mg=0,
          dNTPs=0, saltcorr=5)
```

i.e. The first two arguments are ``seq`` and ``check``, note that ``c_seq`` is the 3rd argument. However, fixing that there is still a ``KeyError``, e.g.

```
>>> import Bio.SeqUtils.MeltingTemp as Tm
>>> Tm.Tm_NN('AAAGCAGCCTAGGACAGCAGT', c_seq='TTTCGTCGGTCCTGTCGTCA', shift=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "Bio/SeqUtils/MeltingTemp.py", line 856, in Tm_NN
    deltaH += de_table[right_de][dH]
KeyError: '.A/GA'
```

@MarkusPiotrowski could you take a look at this please?

**Comment by jennilis** [comment]
Thanks for your reply.  Ah yes, I realized this shortly after posting, both that I was calling it wrong, and that the syntactically correct call still fails, and that is because it is not designed to handle so many mismatches. And I was separately in touch with Markus who was very responsive in pointing out my mistake.

Jennifer

**Comment by MarkusPiotrowski** [comment]
Unfortunately, thermodynamic parameters are not available for all possible base duplets. There are only very few data for double mismatches or mismatches as part of a dangling end duplet. Here, the KeyError is not the result of a bug but the result of missing data.
The best I can do is to give a more precise error message, e.g. throw a ValueError if the key isn't found. This is implemented in the main scanning part of method (lines 924-931) but not (so far) for the dangling ends treatment.

**Comment by peterjc** [comment]
If you can catch this and give a ``ValueError`` instead where the message which captures what you just explained, that would be a nice improvement.

**Comment by snowformatics** [comment]
Hi Jennifer,

would it be possible to help me, I have the same problem. What was the conclusion with Markus?

Thanks!

**Comment by MarkusPiotrowski** [comment]
@snowformatics: This seems indeed to be a bug; looking at `DNA_DE1` and `RNA_DE1` the `key` format for the dangling ends is different...And during dangling ends calculations there is no test for the reverse comlementary key... I will have a look on this.
As workaround you may copy the content of `RNA_DE1` dic and change all keys manually to the correct format, e.g. AA/T. should become .T/AA. Sorry for the inconvenience.

**Comment by snowformatics** [comment]
@MarkusPiotrowski I expected this already and changed the table already. But I was not 100% sure because I got other problems with mismatched combinations as Jennifer. Any hints how to solve that issue? 

BTW, thanks for the amazing documentation!

**Comment by MarkusPiotrowski** [comment]
@snowformatics Thank you. See my comment above from Nov 30. Thermodynamic data are not available for all possible basepair neighbours. Data for double mismatches are almost absent (with a few exceptions, e.g., GG/TT), and there is no data for dangling ends combined with a mismatch (e.g., AA/.C). If such data becomes available, we can include include them.

**Comment by snowformatics** [comment]
Ah ok I see thank you very much!

I am not sure whether this is helpful but here is the corrected table:

RNA_DE2 = {
    '.T/AA': (-4.9, -13.2), '.T/CA': (-0.9, -1.3), '.T/GA': (-5.5, -15.1),
    '.T/AT': (-2.3, -5.5),
    '.G/AC': (-9.0, -23.5), '.G/CC.': (-4.1, -10.6), '.G/GC': (-8.6, -22.2),
    '.G/TC': (-7.5, -20.31),
    '.C/AG': (-7.4, -20.3), '.C/CG': (-2.8, -7.7), '.C/GG': (-6.4, -16.4),
    '.C/TG': (-3.6, -9.7),
    '.T/AG': (-4.9, -13.2), '.T/CG': (-0.9, -1.3), '.T/GG': (-5.5, -15.1),
    '.T/TG': (-2.3, -5.5),
    '.A/AT': (-5.7, -16.1), '.A/CT': (-0.7, -1.9), '.A/GT': (-5.8, -16.4),
    '.A/TT': (-2.2, -6.8),
    '.G/AT': (-5.7, -16.1), '.G/CT': (-0.7, -1.9), '.G/GT': (-5.8, -16.4),
    '.G/TT': (-2.2, -6.8),
    'AT/.A': (-0.5, -0.6), 'CT/.A': (6.9, 22.6), 'GT/.A': (0.6, 2.6),
    'TT/.A': (0.6, 2.6),
    'AG/.C': (-1.6, -4.5), 'CG/.C': (0.7, 3.2), 'GG/.C': (-4.6, -14.8),
    'TG/.C': (-0.4, -1.3),
    'AC/.G': (-2.4, -6.1), 'CC/.G': (3.3, 11.6), 'GC/.G': (0.8, 3.2),
    'TC/.G': (-1.4, -4.2),
    'AT/.G': (-0.5, -0.6), 'CT/.G': (6.9, 22.6), 'GT/.G': (0.6, 2.6),
    'TT/.G': (0.6, 2.6),
    'AA/.T': (1.6, 6.1), 'CA/.T': (2.2, 8.1), 'GA/.T': (0.7, 3.5),
    'TA/.T': (3.1, 10.6),
    'AG/.T': (1.6, 6.1), 'CG/.T': (2.2, 8.1), 'GG/.T': (0.7, 3.5),
    'TG/.T': (3.1, 10.6)}

**Comment by peterjc** [comment]
This should be fixed now via #785, thanks to @MarkusPiotrowski
